src/pc/psxsdk/libspu.c

```diff
@@ -30,9 +30,276 @@ void SpuSetKey(long on_off, unsigned long voice_bit) {
+                        _spu_memList->addr = 0x80001010;
+                        _spu_memList[2].addr = (s32)(temp_a3_2 | 0x40000000);
+                        var_v0_2 = 2;
+                        _spu_memList->size = (u32)temp_a1;
+                        _spu_memList[1].addr = addr_temp;
+                        _spu_memList[1].size = sep_size;
+                        _spu_memList[2].size = size_temp;
+                        goto block_15;
+                    } else {
+                        return -1;
+                    }
+                } else {
+                    var_v0 = -1;
+                    if ((D_8003355C - 1) >= D_80033560) {
+                        _spu_memList[1].addr = (s32)(temp_a3_2 | 0x40000000);
+                        var_v0_2 = 1;
+                        _spu_memList->addr = (u32)addr_temp;
+                        _spu_memList->size = sep_size;
+                        _spu_memList[1].size = size_temp;
+                    block_15:
+                        D_80033560 = var_v0_2;
+                        goto block_30;
+                    }
+                }
+                return var_v0;
+            }
+            func_800286E0();
+            list_pos = 0;
+        loop_18:
+            fmask = 0x0FFFFFFF;
+            fourmask = 0x40000000;
+            if ((s32)_spu_memList[list_pos].addr < 0) {
+                sep_addr = _spu_memList[list_pos].addr & fmask;
+                if ((u32)addr_temp >= sep_addr) {
+                    if (((u32)addr_temp <
+                         (u32)(sep_addr + _spu_memList[list_pos].size)) ||
+                        (_spu_memList[list_pos + 1].addr & fourmask)) {
+                        sep_addr = (u32)addr_temp;
+                        if ((_spu_memList[list_pos].size -
+                             (addr_temp - (_spu_memList[list_pos].addr &
+                                           fmask))) >= (s32)sep_size) {
+                            goto block_29;
+                        }
+                    }
+                    goto block_27;
+                }
+                if ((s32)_spu_memList[list_pos].size < (s32)sep_size) {
+                    list_pos += 1;
+                    goto loop_18;
+                }
+                goto block_29;
+            }
+            if (!(_spu_memList[list_pos].addr & fourmask)) {
+            block_27:
+                list_pos += 1;
+                goto loop_18;
+            }
+            sep_addr = _spu_memList[list_pos].addr & fmask;
+            if ((_spu_memList[list_pos].size - var_s2) >= (s32)sep_size) {
+            block_29:
+                addr_temp = _SpuMallocSeparateTo3(list_pos, sep_addr, sep_size);
+            block_30:
+                func_800286E0();
+                var_v0 = addr_temp;
+                return var_v0;
+            }
+            return -1;
+        }
+    } else {
+        return -1;
+    }
+
+    return var_v0;
+    return 0;
```

Comment:
Is the double-return intentional here?

Reply:
This is just copy-pasted right off my decomp.me scratch, removed